anker/src/logic.rs

```diff
@@ -8,27 +8,62 @@ use solana_program::{
 
 use crate::{instruction::InitializeAccountsInfo, state::Anker};
 
-/// Deserialize the Anker state.
-/// Checks if the Lido instance is the same as the one stored in Anker.
-/// Checks if the Reserve account is the same as the one stored in Anker.
+/// Deserialize the Solido and Anker state.
+///
+/// Check the following things for consistency:
+/// * The Solido state should be owned by the Solido program stored in Anker.
+/// * The Solido state should live at the address stored in Anker.
+/// * The reserve should live at the address derived from Anker.
+/// * The reserve should be valid stSOL account.
+///
+/// The following things are not checked, because these accounts are not always
+/// needed:
+/// * The mint address should match the address stored in Anker.
+/// * The mint authority should match the address derived from Anker.
+/// * The reserve authority should match the address derived from Anker.
 pub fn deserialize_anker(
-    program_id: &Pubkey,
+    anker_program_id: &Pubkey,
     anker_account: &AccountInfo,
-    lido: &Pubkey,
-    reserve_account: &Pubkey,
-) -> Result<Anker, ProgramError> {
-    if anker_account.owner != program_id {
+    solido_account: &AccountInfo,
+    reserve_account: &AccountInfo,
+) -> Result<(Lido, Anker), ProgramError> {
+    if anker_account.owner != anker_program_id {
         msg!(
             "Anker state is owned by {}, but should be owned by the Anker program ({}).",
             anker_account.owner,
-            program_id
+            anker_program_id
         );
-        return Err(LidoError::InvalidOwner.into());
+        return Err(AnkerError::InvalidOwner.into());
     }
+
     let anker = try_from_slice_unchecked::<Anker>(&anker_account.data.borrow())?;
-    anker.check_lido(lido)?;
-    anker.check_reserve_account(program_id, anker_account.key, reserve_account)?;
-    Ok(anker)
+
+    anker.check_self_address(anker_program_id, anker_account)?;
+
+    if *solido_account.owner != anker.solido_program_id {
```

Comment:
I think you are right in this case, but I feel safer checking it anyway. (Remember the Neodyme presentation at Breakpoint? ;-) Also, account owners can change. The Solido instance’s owner can’t because we have no instruction that changes it, but it means that in the future we can end up in a situation where an update to the Solido program makes the existing Anker deployment insecure.